src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```
@@ -0,0 +1,258 @@
+        }
+
+        private static T WrapException<T>(Func<T> func)
+        {
+            try
+            {
+                return func();
+            }
+            catch (DirectoryNotFoundException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (PathTooLongException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (IOException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (NotSupportedException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (SecurityException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (UnauthorizedAccessException ex)
+            {
+                throw CreateException(ex);
+            }
+        }
+
+        private static EgressException CreateException(string message)
+        {
+            return new EgressException(WrapMessage(message));
+        }
+
+        private static EgressException CreateException(Exception innerException)
+        {
+            return new EgressException(WrapMessage(innerException.Message), innerException);
+        }
+
+        private static string WrapMessage(string innerMessage)
+        {
+            return !string.IsNullOrEmpty(innerMessage)
+                ? string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressFileFailedDetailed, innerMessage)
+                : Strings.ErrorMessage_EgressFileFailedGeneric;
+        }
+
+        private static async Task VerifyBucketExistsAsync(IAmazonS3 client, string bucketName, CancellationToken token)
+        {
+            bool exists = await client.DoesS3BucketExistAsync(bucketName);
+            if (exists)
+                return;
+
+            try
+            {
+                await client.PutBucketAsync(bucketName, token);
+            }
+            catch (Exception ex)
```

Do not catch all exceptions. Catch specific ones that will be thrown from here. We don't want to catch exceptions such as `NullReferenceException`, `ObjectDisposedException`, etc that indicate a bug in the code; we only want to catch exceptions that represent an operational failure. This will make it easier for us to determine the cause of the exception when a customer reports that the egress operation fails.